content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt

```diff
@@ -0,0 +1,301 @@
+```
+
+### Result
+If you now add a `(toc)` placeholder to a page that contains headlines of the given level, you should see a ToC appear where you added the placeholder.
+
+## Alternative: A field method to generate the anchors
+
+This is an alternative to using the `kirbyText` hook described above. Do not use the hook and this method in conjuntion.
+
+```php "/site/plugins/toc/index.php"
+<?php
+
+Kirby::plugin('k-cookbook/toc', [
+  'fieldMethods' => [
+    'anchorHeadlines' => function($field, $headlines = 'h2|h3') {
+
+      // create the regex pattern to be used as first argument in `preg_replace_callback()`
+      $headlinesPattern = is_array($headlines) ? implode('|', $headlines) : $headlines;
+
+      // use `preg_replace_callback()` to replace matches with anchors
+      $field->value = preg_replace_callback('!<(' . $headlinesPattern . ')>(.*?)</\\1>!s', function ($match) {
+         
+          // create the id from the headline text
+          $id = Str::slug(Str::unhtml($match[2]));
+
+          // return the modified headline: 
+          // $match[1] contains the match for the first subpattern, i.e. `h2`, `h3` etc.
+          // $match[2] contains the match for the second subpattern, i.e. the actual headline text
+          return '<' . $match[1] . ' id="' . $id . '"><a href="#' . $id . '">' . $match[2] . '</a></' . $match[1] . '>';
+
+      }, $field->value);
+
+      return $field;
+    }
+  ]
+]);
+```
+
+In your templates/snippets, you can now use the field method like this:
+
+### With a single headline level
+
+```php
+<?= $page->text()->kt()->anchorHeadlines('h2') ?>
+```
+
+### With multiple headline levels
+
+```php
+<?= $page->text()->kt()->anchorHeadlines('h2|h3|4') ?>
+```
+
+Instead of passing a string of options separated by `|`, you can also pass an array of options:
+
+```php
+<?= $page->text()->kt()->anchorHeadlines(['h2', 'h3', 'h4']) ?>
+```
+
+If you don't pass an argument, the defaults will be used.
+
+## Alternative: A field method to generate the ToC
+
+```php "/site/plugins/toc/index.php"
+<?php
+
+Kirby::plugin('k-cookbook/toc', [
+  'fieldMethods' => [
+    'headlines' => function($field, $headline = 'h2') {
+
+        preg_match_all('!<' . $headline . '.*?>(.*?)</' . $headline . '>!s', $field->kt()->value(), $matches);
+
+        $headlines = new Collection;
+
+        foreach ($matches[1] as $text) {
+            $headline = new Obj([
+                'id'   => $id = '#' . Str::slug(Str::unhtml($text)),
+                'url'  => $id,
+                'text' => trim(strip_tags($text)),
+            ]);
+
+            $headlines->append($headline->url(), $headline);
+        }
+
+        return $headlines;
+
+    }
+  ]
+]);
+```
+The field method returns a collection of headline objects.
+
+In your templates, you can now use the `headlines()` field method in conjunction with the snippet we registered above.
+
+```php
+<?php snippet('toc', ['headlines' => $page->text()->kt()->headlines('h2')]) ?>
```

Comment:
Doesn’t this need to be this as the method also calls `kt()`?

```suggestion
<?php snippet('toc', ['headlines' => $page->text()->headlines('h2')]) ?>
```

Reply:
Of course!